### PR TITLE
Add Epsilon- and Eta-Sampling

### LINKED ISSUE #21092: Add Epsilon and Eta sampling
### Feature request

I would like to add Epsilon and Eta sampling from [Truncation Sampling as Language Model Desmoothing](https://arxiv.org/abs/2210.15191). They are novel truncation sampling decoding algorithms that have led to better human judgement scores and less repetition than nucleus sampling.

- [Official repository](https://github.com/john-hewitt/truncation-sampling)

### Motivation

I would like to generate more human-like and less repetitious text with Huggingface models.

### Your contribution

I am able to submit a PR for this.

### PR DESCRIPTION
**Pull request by shermansiu**
# What does this PR do?

Implements epsilon- and eta-sampling, as seen in [[Truncation Sampling as Language Model Desmoothing](https://arxiv.org/abs/2210.15191)](https://arxiv.org/abs/2210.15191). The code is adapted from the author's official repository [here](https://github.com/john-hewitt/truncation-sampling).

Resolves #21092.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@gante

### CODE REVIEW

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -323,6 +320,88 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
         return scores
 
 
+class EpsilonLogitsWarper(LogitsWarper):
+    r"""
+    [`LogitsWarper`] that performs epsilon, i.e. restricting to tokens with absolute prob > prob_cut_off. Takes the
```
Comment (gante) [inline src/transformers/generation/logits_process.py:325]:
```suggestion
    [`LogitsWarper`] that performs epsilon-sampling, i.e. restricting to tokens with absolute prob > prob_cut_off. Takes the
```

(you might need to run `make fixup` locally and commit again, after incorporating this change)

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -323,6 +320,88 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+        top_k = min(self.min_tokens_to_keep, scores.size(-1))  # Safety check
+        indices_to_remove = indices_to_remove & (scores < torch.topk(scores, top_k)[0][..., -1, None])
+
+        scores = scores.masked_fill(indices_to_remove, self.filter_value)
+        return scores
+
+
+class EtaLogitsWarper(LogitsWarper):
+    r"""
+    [`LogitsWarper`] that performs eta-sampling. See [Truncation Sampling as Language Model
```
Comment (gante) [inline src/transformers/generation/logits_process.py:368]: Some further description of what eta-sampling is would be appreciated here 🙏

File: src/transformers/generation/configuration_utils.py [hunk 3]
```diff
@@ -109,6 +109,17 @@ class GenerationConfig(PushToHubMixin):
...
             to 1.0 it takes no effect. See [this paper](https://arxiv.org/pdf/2202.00666.pdf) for more details.
+        epsilon_cutoff (`float`, *optional*, defaults to 0):
+            If set to float strictly between 0 and 1, only tokens with a conditional probability greater than
+            `epsilon_cutoff` will be sampled. In the paper, suggested values range from 3e-4 to 9e-4, depending on the
+            size of the model. See [Truncation Sampling as Language Model
+            Desmoothing](https://arxiv.org/abs/2210.15191) for more details.
+        eta_cutoff (`float`, *optional*, defaults to 0):
+            Eta sampling is a hybrid of locally typical sampling and epsilon sampling. If set to float strictly between
+            0 and 1, a token is only considered if it is greater than either the cutoff or sqrt(eta_cutoff) *
+            expected_next_random_token_probability. In the paper, suggested values range from 3e-4 to 2e-3, depending
```
Comment (gante) [inline src/transformers/generation/configuration_utils.py:120]:
```suggestion
            0 and 1, a token is only considered if it is greater than either eta_cutoff or sqrt(eta_cutoff) *
            entropy(softmax(next_token_logits)). In the paper, suggested values range from 3e-4 to 2e-3, depending
```

Adding some precision to the terms here. "expected_next_random_token_probability" isn't exactly what the paper uses, if I got it right :) (feel free to correct me if you think I'm wrong!)

File: src/transformers/generation/utils.py [hunk 4]
```diff
@@ -746,23 +748,22 @@ def _get_logits_warper(
         # all samplers can be found in `generation_utils_samplers.py`
         if generation_config.temperature is not None and generation_config.temperature != 1.0:
             warpers.append(TemperatureLogitsWarper(generation_config.temperature))
+        min_tokens_to_keep = 2 if generation_config.num_beams > 1 else 1
```
Comment (gante) [inline src/transformers/generation/utils.py:751]: Sneaky change, but appreciated! 😉

**Review by gante — COMMENTED** [review]
Thank you for this great contribution! I've added a few comments, but they are very simple to address. 

My suggested timeline for this PR is:
1. I approve this PR after the suggested changes are implemented
2. Sylvain has a look at this PR and approves it
3. We merge it
4. We build a notebook with a few examples and communicate on twitter/LI. If you are interested, we'd be happy to support you here. If you are not, we can take care of it :)

WDYT?

File: src/transformers/generation/configuration_utils.py (hunk 3 above)
Comment (shermansiu) [reply src/transformers/generation/configuration_utils.py:120]: It's actually `exp(-entropy(softmax(next_token_logits)))`, but fixed!

File: src/transformers/generation/logits_process.py (hunk 2 above)
Comment (shermansiu) [reply src/transformers/generation/logits_process.py:368]: Added!

File: src/transformers/configuration_utils.py [hunk 5]
```diff
@@ -144,6 +144,17 @@ class PretrainedConfig(PushToHubMixin):
...
+            If set to float strictly between 0 and 1, only tokens with a conditional probability greater than
+            `epsilon_cutoff` will be sampled. In the paper, suggested values range from 3e-4 to 9e-4, depending on the
+            size of the model. See [Truncation Sampling as Language Model
+            Desmoothing](https://arxiv.org/abs/2210.15191) for more details.
+        eta_cutoff (`float`, *optional*, defaults to 0):
+            Eta sampling is a hybrid of locally typical sampling and epsilon sampling. If set to float strictly between
+            0 and 1, a token is only considered if it is greater than either the cutoff or sqrt(eta_cutoff) *
+            expected_next_random_token_probability. In the paper, suggested values range from 3e-4 to 2e-3, depending
+            on the size of the model. See [Truncation Sampling as Language Model
+            Desmoothing](https://arxiv.org/abs/2210.15191) for more details.
```
[the remark this answers is not in the document]
Comment (shermansiu) [reply src/transformers/configuration_utils.py]: Gotcha! Removed!

File: src/transformers/configuration_utils.py [hunk 6]
```diff
@@ -286,6 +297,8 @@ def __init__(self, **kwargs):
         self.top_k = kwargs.pop("top_k", 50)
         self.top_p = kwargs.pop("top_p", 1.0)
         self.typical_p = kwargs.pop("typical_p", 1.0)
+        self.epsilon_cutoff = kwargs.pop("epsilon_cutoff", 0.0)
+        self.eta_cutoff = kwargs.pop("eta_cutoff", 0.0)
```
[the remark this answers is not in the document]
Comment (shermansiu) [reply src/transformers/configuration_utils.py]: Done!

File: src/transformers/generation/configuration_utils.py [hunk 7]
```diff
@@ -109,6 +109,17 @@ class GenerationConfig(PushToHubMixin):
         typical_p (`float`, *optional*, defaults to 1.0):
             The amount of probability mass from the original distribution to be considered in typical decoding. If set
             to 1.0 it takes no effect. See [this paper](https://arxiv.org/pdf/2202.00666.pdf) for more details.
+        epsilon_cutoff (`float`, *optional*, defaults to 0):
```
[the remark this answers is not in the document]
Comment (shermansiu) [reply src/transformers/generation/configuration_utils.py]: Done!

File: src/transformers/generation/configuration_utils.py [hunk 8]
```diff
@@ -109,6 +109,17 @@ class GenerationConfig(PushToHubMixin):
...
+            If set to float strictly between 0 and 1, only tokens with a conditional probability greater than
+            `epsilon_cutoff` will be sampled. In the paper, suggested values range from 3e-4 to 9e-4, depending on the
+            size of the model. See [Truncation Sampling as Language Model
+            Desmoothing](https://arxiv.org/abs/2210.15191) for more details.
+        eta_cutoff (`float`, *optional*, defaults to 0):
```
[the remark this answers is not in the document]
Comment (shermansiu) [reply src/transformers/generation/configuration_utils.py]: Done!

**Comment by shermansiu** [comment]
@gante I incorporated your changes! I rebased with `main` because there were some merge conflicts, which caused some of the review snippets above to get "outdated," but it's done!

**Review by gante — APPROVED** [review]
LGTM 👍

**Comment by gante** [comment]
@sgugger this PR adds a new sampling-based generation strategy that can be effectively implemented through logits processors

**Review by sgugger — APPROVED** [review]
Thanks a lot for your contribution!

**Comment by gante** [comment]
@shermansiu now step 4 remains :) Would you like to work on it? (we can retweet and share your posts)

**Comment by shermansiu** [comment]
I'm interested, but it feels like it's best added to https://github.com/huggingface/blog/blob/main/notebooks/02_how_to_generate.ipynb than in its own notebook, as other truncation sampling approaches (top-p and top-k) are already there.

**Comment by gante** [comment]
We don't update the notebooks for our blog posts, so they stay consistent with their content :) But we will certainly update our [new docs ](https://huggingface.co/docs/transformers/main/en/generation_strategies)

**Comment by shermansiu** [comment]
I have a lot on my to-do list right now... Although I'd love to contribute to the notebook, I think it's unrealistic for me to be able to put something out soon.

**Comment by shermansiu** [comment]
Yesterday's ACL 2023 tutorial on "Generating Text from Large Language Models" covers eta-sampling and more! John Hewitt, the first author of the eta-sampling paper, was one of the presenters for that tutorial!

Site: https://rycolab.io/classes/acl-2023-tutorial/
Slides: https://drive.google.com/file/d/1UHbGcjzBURG1n2DufC7iDTmGNjIz5Dp_/view
